samples/read_header.c

```diff
@@ -0,0 +1,165 @@
+/** @param argc - count of arguments
+ *  @param argv - pointer to array of arguments
+returns 1 on failure 0 on success
+*/
+int main(int argc, char *argv[])
+{
+    static const struct option lopts[] = {
+        {NULL, 0, NULL, 0}
+    };
+
+    const char *inname = NULL, *tag = NULL, *id = NULL, *opt = NULL;
+    const char *types[] = {"HD", "SQ", "RG", "PG", "CO"}, *idfield[] = { NULL, "SN", "ID", "ID", ""};
+    int c = 0, ret = EXIT_FAILURE, pos = -1, linecnt = 0, headercnt = sizeof(types) / sizeof(types[0]);
+    samFile *infile = NULL;
+    sam_hdr_t *in_samhdr = NULL;
+    kstring_t data = KS_INITIALIZE;
+    const char **header = &types[0], **idf = &idfield[0];
+
+    //[-i infile] [-h header] [-n position] [-t tag] [<infile>]
+    while ((c = getopt_long(argc, argv, "i:t:n:h:", lopts, NULL)) >= 0)
+    {
+        switch (c)
+        {
+        case 'i':                           //infile name
+            inname = optarg;
+            break;
+        case 't':                           //tag
+            tag = optarg;
+            break;
+        case 'h':                           //header
+            opt = optarg;
+            header = &opt;
+            headercnt = 1;
+            //find the id_key field for the requested header type - uses NULL for HD
+            for (c = 0; c < (sizeof(types) / sizeof(types[0])); ++c) {
+                if (!strcmp(types[c], optarg)) {
+                    idf = &idfield[c];
+                    break;
+                }
+            }
+
+            break;
+        case 'n':                           //position
+            pos = atoi(optarg);
+            break;
+        case '?':                           //unknown options found!
+            print_usage(stderr);
+            goto end;
+            break;
+        }
+    }
+
+    if (argc == 1 && isatty(STDIN_FILENO)) {
+        print_usage(stdout);
+        ret = EXIT_SUCCESS;
+        goto end;
+    }
+    //input and output
+    if (!inname && ((argc - optind) == 1))
+        inname = argv[optind];
+    if (!inname)
+        inname = "-";
+
+    //open input files
+    infile = sam_open(inname, "r");
+    if (!infile) {
+        fprintf(stderr, "Could not open %s\n", inname);
+        goto end;
+    }
+
+    //read header
+    in_samhdr = sam_hdr_read(infile);
+    if (!in_samhdr)
+    {
+        fprintf(stderr, "Failed to read header from file!\n");
+        goto end;
+    }
+
+    if (pos == -1) {
+        //position not specified, set to 1st item of the header type
+        pos = 0;
+    }
+
+    if (header == &types[0]) {
+        //no header specified, reset position and tag fields as it would be display of full line
+        tag = NULL;
+        pos = 0;
+    }
+
+    //iterate and show 1st of each type or requested one
+    for (c = 0; headercnt && (c < (sizeof(types) / sizeof(types[0]))); ++c) {
+        //get count of given header type
+        linecnt = sam_hdr_count_lines(in_samhdr, *(header + c));
```

Comment:
`*(header + c)` is better just written as `header[c]` and makes it clearer we're just using an array of header types.